lightning-block-sync/src/lib.rs

```diff
@@ -0,0 +1,1001 @@
+					last_disconnect_tip = Some(header.header.prev_blockhash);
+				},
+				&ForkStep::ForkPoint(ref header) => {
+					new_tip = Some(*header);
+				},
+				_ => {},
+			}
+		}
+
+		// If we disconnected any blocks, we should have new tip data available. If we didn't disconnect
+		// any blocks we shouldn't have set a ForkPoint as there is no fork.
+		assert_eq!(last_disconnect_tip.is_some(), new_tip.is_some());
+		if let &Some(ref tip_header) = &new_tip {
+			debug_assert_eq!(tip_header.header.block_hash(), *last_disconnect_tip.as_ref().unwrap());
+		} else {
+			// Set new_tip to indicate that we got a valid header chain we wanted to connect to, but
+			// failed
+			new_tip = Some(*old_header);
+		}
+
+		for event in events.drain(..).rev() {
+			if let ForkStep::ConnectBlock(header) = event {
+				let block = chain_poller
+					.fetch_block(&header).await
+					.or_else(|e| Err((e, new_tip)))?;
+				debug_assert_eq!(block.block_hash, header.block_hash);
+
+				println!("Connecting block {}", header.block_hash.to_hex());
+				chain_listener.block_connected(&block, header.height);
+				self.header_cache.insert(header.block_hash, header);
+				new_tip = Some(header);
+			}
+		}
+		Ok(())
+	}
+
+	/// Walks backwards from `current_header` and `prev_header`, finding the common ancestor. Returns
+	/// the steps needed to produce the chain with `current_header` as its tip from the chain with
+	/// `prev_header` as its tip. There is no ordering guarantee between different ForkStep types, but
+	/// `DisconnectBlock` and `ConnectBlock` are each returned in height-descending order.
+	async fn find_fork<P: Poll>(&self, current_header: ValidatedBlockHeader, prev_header: &ValidatedBlockHeader, chain_poller: &mut P) -> BlockSourceResult<Vec<ForkStep>> {
+		let mut steps = Vec::new();
+		let mut current = current_header;
+		let mut previous = *prev_header;
+		loop {
+			// Found the parent block.
+			if current.height == previous.height + 1 &&
+					current.header.prev_blockhash == previous.block_hash {
+				steps.push(ForkStep::ConnectBlock(current));
+				break;
+			}
+
+			// Found a chain fork.
+			if current.header.prev_blockhash == previous.header.prev_blockhash {
+				let fork_point = self.look_up_previous_header(chain_poller, &previous).await?;
+				steps.push(ForkStep::DisconnectBlock(previous));
+				steps.push(ForkStep::ConnectBlock(current));
+				steps.push(ForkStep::ForkPoint(fork_point));
+				break;
+			}
+
+			// Walk back the chain, finding blocks needed to connect and disconnect. Only walk back the
+			// header with the greater height, or both if equal heights.
+			let current_height = current.height;
+			let previous_height = previous.height;
+			if current_height <= previous_height {
+				steps.push(ForkStep::DisconnectBlock(previous));
+				previous = self.look_up_previous_header(chain_poller, &previous).await?;
+			}
+			if current_height >= previous_height {
+				steps.push(ForkStep::ConnectBlock(current));
```

Comment:
are the connect block steps guaranteed to be in the correct chronological order? At first glance, working back from the tip, it appears to me as though the connect steps would end up reversed?

Reply:
`sync_listener` will first examine the events in order to process disconnected blocks. Then it will examine the events in reverse order (using `rev()` to process connected blocks.